### PR TITLE
cudaMallocAsync: better VLOG and error message.

### PR DESCRIPTION
**Pull request by nouiz**
This combine a few trivial changes related to memory allocator: 
WAR some OOM, and better VLOG and error message.

In detail:

* cudaMallocAsync:
    - When OOM happens, sync the stream and try to allocate again. This allow extra coalescing/remapping and WAR some OOM.
    - VLOG the peak memory usage.
    - Extra VLOG when OOM including histogram of allocation, the list of allocated ptr and size and cudaMallocAsync stats.
    - Better Error message.
    - Check some return status that wasn't checked.
* cudaMalloc allocator: Add extra VLOG
* BFC Allocator VLOG.
    - Allow to print a trace of allocation/free and to print the peak memory usage.
    - Use increasing VLOG number for basic log, the peak then the trace.

@sanjoy

### CODE REVIEW

**Review by sanjoy — CHANGES_REQUESTED** [review]
Can you please split out the PR into smaller ones?

IMO the following split makes sense:

 - Sync/Retry mechanism
 -  Cosmetic changes: VLOG, error message
 -  Check some return status that wasn't checked

**Comment by nouiz** [comment]
Check the return value: https://github.com/tensorflow/tensorflow/pull/50698
Sync when there is an OOM: https://github.com/tensorflow/tensorflow/pull/50699

And I kept all the cosmetic/  VLOG changes in this PR.

File: tensorflow/core/common_runtime/bfc_allocator.cc [hunk 1]
```diff
@@ -253,7 +255,8 @@ void* BFCAllocator::AllocateRawInternalWithRetry(
 
 void* BFCAllocator::AllocateRaw(size_t unused_alignment, size_t num_bytes,
                                 const AllocationAttributes& allocation_attr) {
-  VLOG(1) << "AllocateRaw " << Name() << "  " << num_bytes;
+  VLOG(3) << "AllocateRaw " << Name() << "  " << num_bytes;
+  void* result = nullptr;
```
Comment (sanjoy) [inline tensorflow/core/common_runtime/bfc_allocator.cc:259]: I'd prefer keeping the previous structure; with this new flow it is easier to forget to set `result` in some branch.

How about using IIFE:

```
void* result = [&] {
  // Old code that does `return AllocateRawInternalWithRetry` etc.
}();
VLOG(3) << ...
return result;
```

File: tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc [hunk 2]
```diff
@@ -214,10 +218,57 @@ void* GpuCudaMallocAsyncAllocator::AllocateRaw(size_t alignment,
...
     mutex_lock lock(lock_);
     LOG(ERROR) << Name() << " cuMemAllocAsync failed to allocate " << num_bytes
-               << ": " << GetCudaErrorMessage(result)
-               << "\n Free memory/Total memory: " << free << "/" << total;
+               << " bytes: " << GetCudaErrorMessage(result)
+               << "\n Reported by CUDA: Free memory/Total memory: " << free << "/" << total;
     if (auto stats = GetStats())
       LOG(ERROR) << "Stats: " << stats->DebugString();
+
+    std::map<size_t, int> size_map_historgram;
```
Comment (sanjoy) [inline tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc:226]: Let's pull all of this out into a function with a clear & obvious name.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc (hunk 2 above)
Comment (nouiz) [reply tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc:226]: done.

File: tensorflow/core/common_runtime/bfc_allocator.cc (hunk 1 above)
Comment (nouiz) [reply tensorflow/core/common_runtime/bfc_allocator.cc:259]: done.

**Review by nouiz — COMMENTED** [review]
Fixed all comments.

**Review by sanjoy — APPROVED** [review]

**Comment by nouiz** [comment]
It didn't got merged. The github Ci status tell me this:
```
feedback/copybara — Google internal checks FAILED for runs with create time 2021-07-16T23:25:25.581986539Z.
```

**Pull request merged**